### DISCUSSION TITLE
Approach to documenting our know-how

### DISCUSSION THREAD

**Discussion by radeusgd** [enso-org/enso#6296] [Work In Progress - Engine]
There are some various opinions about how we should approach documenting the knowledge related to our internal libraries, and there was a proposal to [start a discussion](https://github.com/enso-org/enso/pull/6255#discussion_r1167410742), so I'm starting one.

------

I'd like to discuss how we want to approach 'storing' and searching for information related to our internal libraries - mostly internals of the engine/interpreter, but it can probably also apply to deeper internals of the libs (internal Enso helpers, the polyglot Java libs like Table etc.).

----

- [Java style guide on comments](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md#documentation-comments)

**Comment by radeusgd** [comment]
@JaroslavTulach is suggesting the view that [Pull Requests and Issues are part of the project documentation](https://github.com/enso-org/enso/pull/6255#discussion_r1167390077).

> The **explanation is important**. However I believe consistent explanation of motivations in the PR is even better than random comments spread thru the source code. [Selective Cluelessness](http://wiki.apidesign.org/wiki/Cluelessness) asks for making a _"knowledge increasable"_. Should the code not be self explanatory, then all one needs to do to use `git blame` and locate the #6255 PR.
>
> (...) Duplicating randomly identified parts of such description in the source code makes little sense. It'd be even more misleading than useful.
>
> Let's treat **Pull Requests and Issues as natural part of the project**. Let's learn how to locate them when seeing incomprehensible source code.

**Reply by radeusgd** [reply]
> > (...) Duplicating randomly identified parts of such description in the source code makes little sense. It'd be even more misleading than useful.

@JaroslavTulach why do you think that adding a _summary/explanation_ based on a PR discussion into the code itself would be misleading?

> > Let's treat **Pull Requests and Issues as natural part of the project**. Let's learn how to locate them when seeing incomprehensible source code.

This is complicating understanding a piece of code **a lot**. Let's say I'm reading a class trying to add some functionality to it, or doing code review on the PR. I need to understand how it works. IMO I should have all necessary explanations directly in the code or linked from it. If I need to start doing `git blame` on each file I try to understand, in hope that _maybe there is some comment in some ancient PR explaining it_ - I'm becoming an archaeologist. `blame` for one file may open up 10s of commits, since the files tend to evolve a lot. The last commit in which a line was modified may not necessarily be the right one, because the explanation may have appeared in an earlier modification. So one has to essentially perform a whole investigation to find all the PRs that touched the file and search if any of it contains some discussion that may shed some light on what a thing is supposed to do and why. 

I can't imagine such approach as being viable. Of course, maybe if I need to understand some obscure thing, or I'm new - I may need to do more searching. At this point it's probably better to just ask other devs which know the files well and get some human to explain it. 

But if something is not obvious to developers knowing the codebase, I think it warrants putting a note in the code. Making developers regularly search through history is turning them into archivists.

**Reply by JaroslavTulach** [reply]
Just before joining Enso I was coding for the Oracle DB - the flagship product that keeps Oracle operating for almost fifty years. The first thing one has to do before coding anything is to do a _marketing document_, if that gets approved one can start writing _functional specification_, based on that another team writes _test specification_. When all is approved, one can start coding. The _code itself isn't its only documentation_, **all the documents compose it** and all _keeps together via commit messages and bugtracking system_. 

> > > Let's treat **Pull Requests and Issues as natural part of the project**. 

Just to be sure: Is there anyone disputing that issue tracker is a natural part of the project? If not, then it is just about: _Where do we draw the line_? How much do we document in design **specifications**, how much do we put into the **issues** and how much do we put into the **code**?

> summary/explanation based on a PR discussion into the code itself would be misleading

Because the [full analysis of the problem including description of the solution](https://github.com/enso-org/enso/pull/6255#issuecomment-1506293252) is in my PR. It needs to be somewhere **as a whole**, extracting random parts of it makes no sense. We can indeed create a specification document and put there the same - but we are 
startup, we aren't coding _"OracleDB"_ yet - e.g. maybe requiring creation of a design document for every fix is a bit too much?

> > > Let's learn how to locate them when seeing incomprehensible source code.

When we decide a design document is an overkill for some bugfix, then **issue tracker is a great place to summarize and describe** the overall change. 

Bugtracker and commits are _the glue that holds the history of any project together_. But we have to learn how to use it: Take a look [here](https://github.com/enso-org/enso/pull/5998#discussion_r1167633535) - I'd expect a reference to commit that _"done it"_ - without it things **aren't glued and can't be re-understand** easily in the future. And that is important as ...

> I'm becoming an archaeologist

...after a while **every developer** of a successful project turns into an **archaeologist**. We should _realize it now_ and make sure future _archaeologist activities_ will provide _pleasant_ experience!

**Reply by Akirathan** [reply]
Let me contribute to the discussion. I am going to structure my message into two sections; the first one describes what I believe is the best possible approach towards documentation and code, and the second one replies to the comments above.

The reason why the motivation section is so long is that I disagree with the approach that we should refactor code only if it increases performance or if it is part of the public API.

# My goal and my motivation
I think that we can all agree that issues, pull request, discussion, etc. are all relevant part of the project. However, what I am advocating for (and I believe @radeusgd is also advocating for) is a [**Clean Code**](https://www.goodreads.com/book/show/3735293-clean-code).

There is a great and inspiring book by Robert C. Martin about what he calls *clean code*. In the very first introductory chapters, there are many quotations from very well known and respected authorities in the software engineering field, like Bjarne Stroustrup, Grady Booch, Dave Thomas, and others.  And to all my understanding, what they all agree on is that **the cost of a bad code is huge**. The following two paragraphs, along with the figure, are quotations from that book, which I think stress out my point:

------------------------

## The total cost of owning a mess
If you have been a programmer for more than two or three years, you have probably been signiﬁcantly slowed down by messy code. The degree of the slowdown can be signiﬁcant. Over the span of a year or two, teams that were moving very fast at the beginning of a project can ﬁnd themselves moving at a snail’s pace. Every change they make to the code breaks two or three other parts of the code. No change is trivial. Every addition or modiﬁcation to the system requires that the tangles, twists, and knots be “understood” so that more tangles, twists, and knots can be added. Over time the mess becomes so big and so deep and so tall, they can not clean it up. There is no way at all.

As the mess builds, the productivity of the team continues to decrease, asymptotically approaching zero. As productivity decreases, management does the only thing they can; they add more staff to the project in hopes of increasing productivity. But that new staff is
not versed in the design of the system. They don’t know the difference between a change that matches the design intent and a change that thwarts the design intent. Furthermore, they, and everyone else on the team, are under horrific pressure to increase productivity. So they all make more and more messes, driving the productivity ever further toward zero:

![Screenshot from 2023-04-16 14-16-26](https://user-images.githubusercontent.com/14013887/232309364-e23050e4-7142-4115-80f2-7ab734d964b5.png)

## Why clean code matters?
The ratio of time spent reading vs. writing code is well over 10:1. We are constantly reading old code as part of the effort to write new code. Because this ratio is so high, we want the reading of code to be easy, even if it makes the writing harder. Of course, there’s no way to write code without reading it, so making it *easy to read actually makes it easier to write.

------------------------

So I claim that taking a little extra effort to comply with some best practices in the field, which often require a bit of refactoring like identifier renaming, splitting code into multiple methods, etc., will pay off heavily in the future, as it means that time to read, and understand the code, will decrease.

Since we all strive to make Enso a good, usable, and stable product, we should spend some effort in order to make the code base as readable and understandable as possible, so that the maintenance cost in the future is minimized.

# Documentation in VCS?
Now, back to @JaroslavTulach 's comment about the information in VCS. I agree that from time to time, we have to look into some older PRs in order to properly understand the motivation of some code (one particular use case that I can think of now is a removal of some old unused code). But I don't see a reason why should I do that in order to understand a piece of code that could have been refactored, and documented, right away. Even in our Java code style guide, we have a [Source Notes section](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md#source-notes) that says that if a piece of code might be too difficult to understand for any reasons, a documentation should be written in the source code.

**Reply by radeusgd** [reply]
I agree with the above.

To sum up and partially reply - I think that yeah in some cases we do need to 'become an archealogist' - if we want to understand some more complex motivation of a thing. I guess that is fine.

But when I look at a piece of code - I shouldn't need to perform archaeology to understand how it works. Adding comments or renaming methods to be more readable is not duplicating random parts of the design.

I think our documentation is lacking and this will bite us but at the same time I'm not proposing having to write lengthy design documents for every fix - I'm not sure what is the best solution here. Maybe we should try to have some doc summing up the overall design directions and try to update it from time to time? I think it may prove to be useful at some point, but I'm not pushing for it.

I feel like maybe we misunderstood on the PR too?

> Because the https://github.com/enso-org/enso/pull/6255#issuecomment-1506293252 is in my PR. It needs to be somewhere as a whole,

True! But what I think me and @Akirathan meant is not necessarily to describe in detail _why_ we are sometimes adding frames and why we introduced the inlining. The request was to document the fact that 'having a `VirtualFrame` argument, influences if the builtin will be inlined or not'. This is a functionality that is there, but it to understand it one has to dig through the code quite a bit, while a short note on `needsFrame` could help it. 

Also, to understand the motivations for inlining - a link to the PR in a doc comment could also greatly help in the search. 

Is there anything problematic which adding links to other pieces of our documentation? Then we know there are resources to look into. 

I imagine when introducing a change, I'm not capable to dig through all relevant PRs for every little change - this would take too much time. Especially, as in most cases I will _not get much useful information this way anyway_. So having links/notes in the relevant places helps figure out when it is worth to do the search.

**Reply by JaroslavTulach** [reply]
> Clean code

Clean code Is indeed desirable. Remaining issue: What methodologies lead us to the clean code?

My favorite one is [TDD](https://en.wikipedia.org/wiki/Test-driven_development) and particularly its _"Write the simplest code that passes the new test"_. Hence my objection against
- #6293

I feel there is no justification to implement the more complex solution _right now_ when the `VirtualFrame` check is good enough. Hopefully the documentation problem with `VirtualFrame` is now addressed by [improvements to Javadoc](https://github.com/orgs/enso-org/discussions/6296#discussioncomment-5644666).

**Reply by JaroslavTulach** [reply]
Few self references from my past:

> slowed down by messy code ... the mess becomes so big

That frequently happens. I call it the *entropy of software*, page 60 of the [Practical API Design](https://theswissbay.ch/pdf/Gentoomen%20Library/Programming/Misc/Apress.Practical.API.Design.Confessions.of.a.Java.Framework.Architect.Jul.2008.pdf).  One of the best tools to prevent it is to read [modularity manifesto](http://wiki.apidesign.org/wiki/Modularity). I wish the Enso code was more modular and enforced non-cyclic references more.

> difference between a change that matches the design intent

My take on [quality](http://wiki.apidesign.org/wiki/Quality). Also available on page 49 of the [Practical API Design](https://theswissbay.ch/pdf/Gentoomen%20Library/Programming/Misc/Apress.Practical.API.Design.Confessions.of.a.Java.Framework.Architect.Jul.2008.pdf). 

> constantly reading old code as part of the effort to write new code

Alas that's what happens to every successful project that lasts years. That's why I'd like us to [get ready for becoming archeologist now](https://github.com/orgs/enso-org/discussions/6296#discussioncomment-5626754) and always leave tons of _"decision evidence"_ in VCS and issue tracker.

**Reply by radeusgd** [reply]
The Javadoc addition looks perfect - that's the kind of comment that I thought would help us, thanks!

I guess you are right that sometimes we just have to be archeologists - and I have nothing against leaving decision evidence in the VCS and issues - I think it's a brilliant idea.

I just thought that we should not need to dig very deep to understand _not decisions but current implementation_. Of course a balance is needed here. Having docs like this added Javadoc is IMO the right balance between too much and too little documentation.

**Comment by radeusgd** [comment]
In my opinion, the primary place for documentation of _code_ is _code itself_.

Ideally, every non-obvious behaviour of a method (one that is not apparent from the method name and type), should get a note in the method docstring. And every class whose purpose is not obvious should have a note explaining it.

If there is a decision that is worth explaining, ideally I think it is good to add a short note about it in the code. IMO it makes it easiest to read the code - because when I look at it, I immediately see any related explanations. They should not be pages long - just short notes explaining the intent or reasons to do something. They should not state the obvious, or stuff that is common knowledge in the team - they should be used for places where it was not clear. If there was a discussion on a PR how to do something - IMO such a discussion suggests that something was not clear and a short note about it is a good idea.

The code should be readable and express the intent well and thus need little comments. But sometimes it is not possible due to performance restrictions, or just complex logic - in that case some explanation what is happening helps immensely to understand what was the intention of the person writing it. _Some examples from our libraries: [(1)](https://github.com/enso-org/enso/blob/develop/distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Dialect.enso#L147), [(2)](https://github.com/enso-org/enso/blob/develop/distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Postgres_Type_Mapping.enso#L23), [(3)](https://github.com/enso-org/enso/blob/develop/distribution/lib/Standard/Database/0.0.0-dev/src/Internal/SQLite/SQLite_Type_Mapping.enso#L20-L50), [(4)](https://github.com/enso-org/enso/blob/develop/std-bits/table/src/main/java/org/enso/table/data/column/storage/type/StorageType.java#L3-L5), [(5)](https://github.com/enso-org/enso/blob/develop/std-bits/table/src/main/java/org/enso/table/data/column/storage/MixedStorageFacade.java#L13-L18). And for example from @hubertp's code: [(6)](https://github.com/enso-org/enso/blob/b8ceaeccdcabaa7aa7913c89fbdcfb6bf4902292/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/TypeProcessor.java#L55-L57), [(7)](https://github.com/enso-org/enso/blame/b8ceaeccdcabaa7aa7913c89fbdcfb6bf4902292/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/builtins/MethodGenerator.java#L127-L137)._

For more complex concepts, spanning across many files or projects, we have the `docs` directory where we can explain the overall architecture. This directory is still updated, but I feel like it is getting much less traffic than some time ago. _For example, [Function call flow](https://github.com/enso-org/enso/blob/develop/docs/runtime/function-call-flow.md) beautifully explains how our function calling works - I'm worried how up-to-date that is, but back in the day it helped me understand this architecture - which was a bit complex. Then we have this awesome docs of the [annotations for builtins](https://github.com/enso-org/enso/blob/b8ceaeccdcabaa7aa7913c89fbdcfb6bf4902292/lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/Builtin.java)._

I definitely don't want to flood us with documentation, but I think we are lacking it in some places, making onboarding of new developers potentially harder and worsening our '[bus factor](https://en.wikipedia.org/wiki/Bus_factor)'.

**Reply by JaroslavTulach** [reply]
>  place for documentation of code is code itself.

Some methodologies advocate _self-documenting code_ and some (extreme) methodologies even prevent comments in the code claiming that then the identifier names, function names & etc. get better as a compensation for missing `// comments`.

In any case Java also offers  **Javadoc**. That is a standard and suitable place for (even internal) API documentation. Hence I put the [explanation of the VirtualFrame argument](https://github.com/enso-org/enso/pull/6304/files#diff-742a6a9fd5466c9a92a8f2f5d969c75e44c7e4869b2b0a47f03a362682b0df7bR11) behavior into Javadoc of `@BuiltinMethod` annotation. IDEs shall offer you the text when using/hovering over the `@BuiltinMethod` usage.

![Usage in EqualsNode](https://user-images.githubusercontent.com/26887752/232676226-fe6bf37f-81c2-4a74-81c1-045d53b6a446.png)

Having the documentation visible at _usage site_ is clearly a practical advantage of having good Javadoc. Writing Javadoc is something I can only encourage in spite of belonging more to the _self-documenting code_ camp.
